src/CloudNative.CloudEvents.Amqp/AmqpClientExtensions.cs

```diff
@@ -94,5 +95,96 @@ private static bool HasCloudEventsContentType(Message message, out string conten
+                        Binary = formatter.EncodeStructuredModeMessage(cloudEvent, out var contentType)
+                    };
+                    // TODO: What about the other parts of the content type?
+                    properties = new Properties { ContentType = contentType.MediaType };
+                    break;
+                case ContentMode.Binary:
+                    bodySection = SerializeData(cloudEvent.Data);
+                    properties = new Properties { ContentType = cloudEvent.DataContentType };
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(contentMode), $"Unsupported content mode: {contentMode}");
+            }
+            return new Message
+            {
+                ApplicationProperties = applicationProperties,
+                BodySection = bodySection,
+                Properties = properties
+            };
+        }
+
+        private static ApplicationProperties MapHeaders(CloudEvent cloudEvent)
+        {
+            var applicationProperties = new ApplicationProperties();
+            var properties = applicationProperties.Map;
+            properties.Add(SpecVersionAmqpHeader, cloudEvent.SpecVersion.VersionId);
+
+            foreach (var pair in cloudEvent.GetPopulatedAttributes())
+            {
+                var attribute = pair.Key;
+
+                // The content type is specified elsewhere.
+                if (attribute == cloudEvent.SpecVersion.DataContentTypeAttribute)
+                {
+                    continue;
+                }
+
+                string propKey = AmqpHeaderPrefix + attribute.Name;
+
+                // TODO: Check that AMQP can handle byte[], bool and int values
+                object propValue = pair.Value switch
+                {
+                    Uri uri => uri.ToString(),
+                    // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
+                    // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
+                    DateTimeOffset dto => dto.UtcDateTime,
+                    _ => pair.Value
+                };
+                properties.Add(propKey, propValue);
+            }
+            return applicationProperties;
+        }
+
+        /// <summary>
+        /// Convert data into a suitable format for inclusion in an Avro record.
+        /// TODO: Asynchronous version?
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        private static RestrictedDescribed SerializeData(object data)
+        {
+            switch (data)
+            {
+                case null:
+                    return null;
+                case byte[] bytes:
+                    return new Data { Binary = bytes };
+                case MemoryStream memoryStream:
+                    // Note: this will return the whole stream, regardless of position...
```

Comment:
Hmm, I hadn't noticed this before. Why don't we want to respect the position?

Reply:
Indeed. I'm pretty dubious about supporting `Stream` as a value in `CloudEvent.Data` at all, to be honest. It feels really fragile - who is responsible for closing the stream?

This PR doesn't change the current behavior though, it just moves it.

I've filed #103 to re-examine this.